test/test_tools.py

```diff
@@ -25,6 +24,108 @@
+import pytest
+import numpy as np
+import pyopencl as cl
+import pyopencl.array  # noqa
+from pyopencl.tools import (  # noqa
+        pytest_generate_tests_for_pyopencl as pytest_generate_tests)
+
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.parametrize("p", [1, 2, 3, 4, 5, 6, 7, 8, 16, 17])
+def test_allreduce_comm_pattern(p):
+    from boxtree.tools import AllReduceCommPattern
+
+    # This models the parallel allreduce communication pattern.
+
+    # processor -> communication pattern of the processor
+    patterns = [AllReduceCommPattern(i, p) for i in range(p)]
+    # processor -> list of data items on the processor
+    data = [[i] for i in range(p)]
+    from copy import deepcopy
+
+    while not all(pat.done() for pat in patterns):
+        new_data = deepcopy(data)
+
+        for i in range(p):
+            if patterns[i].done():
+                for pat in patterns:
+                    if not pat.done():
+                        assert i not in pat.sources() | pat.sinks()
+                continue
+
+            # Check sources / sinks match up
+            for s in patterns[i].sinks():
+                assert i in patterns[s].sources()
+
+            for s in patterns[i].sources():
+                assert i in patterns[s].sinks()
+
+            # Send / recv data
+            for s in patterns[i].sinks():
+                new_data[s].extend(data[i])
+
+        for pat in patterns:
+            if not pat.done():
+                pat.advance()
+        data = new_data
+
+    for item in data:
+        assert len(item) == p
+        assert set(item) == set(range(p))
+
+
+@pytest.mark.parametrize("order", "CF")
+def test_masked_matrix_compression(ctx_getter, order):
+    cl_context = ctx_getter()
+
+    from boxtree.tools import MaskCompressorKernel
+    matcompr = MaskCompressorKernel(cl_context)
+
+    n = 40
+    m = 10
+
+    np.random.seed(15)
+
+    arr = (np.random.rand(n, m) > 0.5).astype(np.int8).copy(order=order)
+
+    with cl.CommandQueue(cl_context) as q:
+        d_arr = cl.array.Array(q, (n, m), arr.dtype, order=order)
+        d_arr[:] = arr
+        arr_starts, arr_lists, evt = matcompr(q, d_arr)
+        cl.wait_for_events([evt])
+        arr_starts = arr_starts.get(q)
+        arr_lists = arr_lists.get(q)
+
+    for i in range(n):
+        items = arr_lists[arr_starts[i]:arr_starts[i+1]]
+        assert set(items) == set(arr[i].nonzero()[0])
+
+
+def test_masked_list_compression(ctx_getter):
+    cl_context = ctx_getter()
+
+    from boxtree.tools import MaskCompressorKernel
+    listcompr = MaskCompressorKernel(cl_context)
+
+    n = 20
+
+    np.random.seed(15)
+
+    arr = (np.random.rand(n) > 0.5).astype(np.int8)
+
+    with cl.CommandQueue(cl_context) as q:
+        d_arr = cl.array.to_device(q, arr)
+        arr_list, evt = listcompr(q, d_arr)
+        cl.wait_for_events([evt])
+        arr_list = arr_list.get(q)
+
+    assert set(arr_list) == set(arr.nonzero()[0])
+
 
 import pytest
```

Comment:
Imports should stay at the top of the file.

Reply:
`pytest` is already imported, so removed here.